Add opt-in LogRecordProcessor.Enabled (#4439)
open-telemetry/opentelemetry-specification: specification/logs/sdk.md

the `OnEmit` call. If `logRecord` is needed after `OnEmit` returns (i.e. for
asynchronous processing) only reads are permitted.

#### Enabled

**Status**: [Development](../document-status.md)

`Enabled` is an operation that a `LogRecordProcessor` MAY implement

> in order to support filtering via [`Logger.Enabled`](api.md#enabled).

Reviewer: Not sure where to mention this , but it is important to call out that, if a EnrichmentProcessor is used, it must be used as a wrapper around the exporting processor to benefit from the perf gains of Enabled() check. 
Otherwise, EnrichmentProcessor would return true for Enabled(), forcing the Logger to anyway create LogRecord and invoke OnEmit(), which gets Enriched by EnrichmentProcessor, and then passed to Exporting processor which would just drop it.

its covered in the supplementary guidelines example, but wondering if we should make it more obvious in the spec itself?

Author: > its covered in the supplementary guidelines example, but wondering if we should make it more obvious in the spec itself?

I find supplementary guidelines are currently the best place so that we do not make the specification too verbose. I think we can also try refining it in a separate PR unless you have some nice proposal how to describe it here.

**Parameters:**

> * [Instrumentation Scope](./data-model.md#field-instrumentationscope) associated
>   with the `Logger`
> * [Context](../context/README.md) explicitly passed by the caller or the current
>   Context

Reviewer 2: ```suggestion
* [Context](../context/README.md) explicitly passed by the caller or the current
  Context
* [Instrumentation Scope](./data-model.md#field-instrumentationscope) associated
  with the `Logger`
```

nit. swap the order

* [Severity Number](./data-model.md#field-severitynumber) passed by the caller

**Returns:** `Boolean`

> Any modifications to parameters inside `Enabled` MUST NOT be propagated to the

Reviewer: Nit: is there a need to specify this? I don't think we do that in other places of the spec.

Author: Not sure. I prefered to call it out explicitly.

Reviewer: no harm keeping it.

caller. Parameters are immutable or passed by value.

An implementation should default to returning `true` for an indeterminate

> state, but may return `false` if valid reasons in particular circumstances

Reviewer 3: would we consider implementing `Enabled` on Batching Processor (and it'd return false when queue is full)?

Author: I think that an SDK can do that. However, I am not sure if it is worth doing, because:
1. it is not something that is on the hot-path (it should not happen frequently)
2. it could "race", in a few CPU cycles later the queue may no longer be full

> An implementation should default to returning `true` for an indeterminate
> state, but may return `false` if valid reasons in particular circumstances
> exist (e.g. performance).

Reviewer 3: I think I understand the intention, maybe we can rephrase it to be more clear? E.g.

```suggestion
`LogRecordProcessor` implementations SHOULD support this operation if they intend to filter out `LogRecord`s. An implementation should return `false` when it's going to filter out the corresponding `LogRecord` out (if the record is ever created). It SHOULD default to returning `true` for any indeterminate state, for example, when awaiting configuration to be provided.
```

Currently it feels like processor can just decide to return false for reasons not necessarily related to its main functionality.

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4439/commits/4f60c059b566eac83c490f35290eae3a4ed644ec

> `LogRecordProcessor` implementations that choose to support this operation are
> expected to re-evaluate the [ReadWriteLogRecord](#readwritelogrecord) passed to

Reviewer 3: looking at the `IsolatedProcessor` example, it's only the case when the processor does some filtering (and not the case when processor wraps others that might do the filtering). Perhaps something along these lines would be more precise?

```suggestion
`LogRecordProcessor` implementations that filter out some log records in `OnEmit` callback and
choose to support `Enable` operation are expected to re-evaluate the [ReadWriteLogRecord](#readwritelogrecord) passed to
```

Author: Indeed. nice and important catch

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4439/commits/4f60c059b566eac83c490f35290eae3a4ed644ec

> [`OnEmit`](#onemit). It is not expected that the caller to `OnEmit` will use the

Reviewer: Why is it that the caller is _**not expected**_ to skip calling `OnEmit`? Is it to account for the multi-processors scenario, where some of then return true and some false, so caller will call OnEmit anyway?

the `OnEmit` call. If `logRecord` is needed after `OnEmit` returns (i.e. for
asynchronous processing) only reads are permitted.

#### Enabled

**Status**: [Development](../document-status.md)

> `Enabled` is an opt-in operation that a `LogRecordProcessor` may implement

Reviewer 4: Does "may" here need to be a key-word?

```suggestion
`Enabled` is an opt-in operation that a `LogRecordProcessor` MAY implement
```

> `Enabled` is an opt-in operation that a `LogRecordProcessor` MAY implement

Reviewer 3: what does opt-in mean, could you please elaborate?

Author: It means that it not must implemented. I think it is redundant with 

> that a `LogRecordProcessor` MAY implement in order to support filtering via [`Logger.Enabled`](api.md#enabled).

Probably removing opt-in will be less confusing. WDYT?

```suggestion
`Enabled` is an operation that a `LogRecordProcessor` MAY implement
```

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4439/commits/2ff3f5a2870cab1431bbe20f99bb2e490fb2f352

in order to support filtering via [`Logger.Enabled`](api.md#enabled).

**Parameters:**

> * [Resource](./data-model.md#field-resource) associated with the `LoggerProvider`

Reviewer: Do we have to pass Resource on each call? It should be possible to do it more efficiently. I suggest to remove it or mention that Resource information is available to LogProcessor. Or remove it completely.

Author: @cijothomas, you mean not even even passing a reference/pointer to the resource, right?

"Resource information is available to LogProcessor" seems fuzzy/unclear to me.

Personally, I am OK with removing it. We can always add it later. I had not experienced any use case where the resource information has any information needed for filtering.

@MrAlias, do you have any preferences? I am asking as you already approved and maybe you have some opinion.

Author: ```suggestion
```

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4439/commits/df9ebb1014156de3254bfb80fa979b82f6d2fcbe

Reviewer 4: > @MrAlias, do you have any preferences? I am asking as you already approved and maybe you have some opinion.

SGTM

the `OnEmit` call. If `logRecord` is needed after `OnEmit` returns (i.e. for
asynchronous processing) only reads are permitted.

#### Enabled

**Status**: [Development](../document-status.md)

`Enabled` is an operation that a `LogRecordProcessor` MAY implement
in order to support filtering via [`Logger.Enabled`](api.md#enabled).

**Parameters:**

* [Context](../context/README.md) explicitly passed by the caller or the current
  Context
* [Instrumentation Scope](./data-model.md#field-instrumentationscope) associated
  with the `Logger`
* [Severity Number](./data-model.md#field-severitynumber) passed by the caller

**Returns:** `Boolean`

An implementation should return `false` if a `LogRecord` (if ever created)
is supposed to be filtered out for the given parameters.
It should default to returning `true` for any indeterminate state, for example,
when awaiting configuration.

Any modifications to parameters inside `Enabled` MUST NOT be propagated to the
caller. Parameters are immutable or passed by value.

> This operation is usually called synchronously, therefore it should not block

Reviewer 5: Who calls this, and when?

Author: It is already described in the spec.
It can be called by the `Logger.Enabled` implementation of the SDK and a processor which is wrapping other processor.

Reviewer 5: Oh I missed the line above:

> - all registered `LogRecordProcessors` implement [`Enabled`](#enabled-1), and a call to `Enabled` on each of them returns `false`.

My mistake 🙂
